parser-typechecker/src/Unison/Lexer.hs

```diff
@@ -60,74 +97,440 @@ type IsVirtual = Bool -- is it a virtual semi or an actual semi?
 
 makePrisms ''Lexeme
 
+space :: P ()
+space = LP.space CP.space1 (fold <|> LP.skipLineComment "--")
+                           (LP.skipBlockCommentNested "{-" "-}")
+  where
+  fold = P.try $ lit "---" *> P.takeRest *> pure ()
+
+lit :: String -> P String
+lit = P.try . LP.symbol (pure ())
+
+token :: P Lexeme -> P [Token Lexeme]
+token = token' (\a start end -> [Token a start end])
+
+pos :: P Pos
+pos = do
+  p <- P.getPosition
+  pure $ Pos (P.unPos (P.sourceLine p)) (P.unPos (P.sourceColumn p))
+
+-- Token parser: strips trailing whitespace and comments after a
+-- successful parse, and also takes care of emitting layout tokens
+-- (such as virtual semicolons and closing tokens).
+token' :: (a -> Pos -> Pos -> [Token Lexeme]) -> P a -> P [Token Lexeme]
+token' tok p = LP.lexeme space (token'' tok p)
+
+-- Committed failure
+err :: Pos -> Err -> P x
+err start t = do
+  stop <- pos
+  _ <- void CP.anyChar <|> P.eof
+  P.customFailure (Token t start stop)
+
+{-
+commitAfter :: P a -> (a -> P b) -> P b
+commitAfter a f = do
+  a <- P.try a
+  f a
+-}
+
+commitAfter2 :: P a -> P b -> (a -> b -> P c) -> P c
+commitAfter2 a b f = do
+  (a,b) <- P.try $ liftA2 (,) a b
+  f a b
+
+-- Token parser implementation which leaves trailing whitespace and comments
+-- but does emit layout tokens such as virtual semicolons and closing tokens.
+token'' :: (a -> Pos -> Pos -> [Token Lexeme]) -> P a -> P [Token Lexeme]
+token'' tok p = do
+  start <- pos
+  -- We save the current state so we can backtrack the state if `p` fails.
+  env <- S.get
+  layoutToks <- case opening env of
+    -- If we're opening a block of type b, we push (b, currentColumn) onto
```

Comment:
Here again is a reference to the "block of type b" that looks like it was monomorphised to "String" above. Should this comment be adjusted too or does it still make sense?

Reply:
block type just meant like "handle" or "if", that sort of thing, not like the static type of the block. Maybe I'll try to think of a better wording. :)

Reply:
came up with maybe better wording, said "block name" instead of "block type"